adapters/socketio_berkeley.c

```diff
@@ -659,6 +500,220 @@ static int set_target_network_interface(int socket, char* mac_address)
+            result = MU_FAILURE;
+        }
+        else
+        {
+            addr = dns_resolver_get_addrInfo(socket_io_instance->dns_resolver);
+
+            if (addr == NULL)
+            {
+                LogError("DNS resolution failed");
+                result = MU_FAILURE;
+            }
+#ifdef IPV6_ENABLED
+            else if (addr->ai_family == AF_INET6)
+            {
+                connect_addr6 = (struct sockaddr_in6*) addr->ai_addr;
+                connect_addr_len = sizeof(struct sockaddr_in6);
+                result = 0;
+            }
+#endif // IPV6_ENABLED
+            else
+            {
+                connect_addr = addr->ai_addr;
+                connect_addr_len = sizeof(*addr->ai_addr);
+                result = 0;
+            }
+        }
+    }
+    else
+    {
+        size_t hostname_len = strlen(socket_io_instance->hostname);
+        if (hostname_len + 1 > sizeof(addrInfoUn.sun_path))
+        {
+            LogError("Hostname %s is too long for a unix socket (max len = %lu)", socket_io_instance->hostname, (unsigned long)sizeof(addrInfoUn.sun_path));
+            result = MU_FAILURE;
+        }
+        else
+        {
+            memset(&addrInfoUn, 0, sizeof(addrInfoUn));
+            addrInfoUn.sun_family = AF_UNIX;
+            // No need to add NULL terminator due to the above memset
+            (void)memcpy(addrInfoUn.sun_path, socket_io_instance->hostname, hostname_len);
+
+            connect_addr = (struct sockaddr*)&addrInfoUn;
+            connect_addr_len = sizeof(addrInfoUn);
+            result = 0;
+        }
+    }
+
+    socket_io_instance->socket = socket (socket_io_instance->address_type == ADDRESS_TYPE_IP ? addr->ai_family : AF_UNIX, SOCK_STREAM, 0);
+
+    if (socket_io_instance->socket < SOCKET_SUCCESS)
+    {
+        LogError("Failure: socket create failure %d.", socket_io_instance->socket);
+        result = MU_FAILURE;
+    }
+#ifndef __APPLE__
+    else if (socket_io_instance->target_mac_address != NULL &&
+                set_target_network_interface(socket_io_instance->socket, socket_io_instance->target_mac_address) != 0)
+    {
+        LogError("Failure: failed selecting target network interface (MACADDR=%s).", socket_io_instance->target_mac_address);
+        result = MU_FAILURE;
+    }
+#endif //__APPLE__
+
+    if(result == 0)
+    {
+        if ((-1 == (flags = fcntl(socket_io_instance->socket, F_GETFL, 0))) ||
+            (fcntl(socket_io_instance->socket, F_SETFL, flags | O_NONBLOCK) == -1))
+        {
+            LogError("Failure: fcntl failure.");
+            result = MU_FAILURE;
+        }
+        else
+        {
+#ifndef IPV6_ENABLED
+            result = connect(socket_io_instance->socket, connect_addr, connect_addr_len);
+#else
+            if (addr->ai_family == AF_INET6)
+            {
+                result = connect(socket_io_instance->socket, (struct sockaddr*)connect_addr6, connect_addr_len);
```

Comment:
This should be the same connect, using the DNS record's ai_addr and ai_addrlen. Please remove the cast.

Reply:
See https://gist.github.com/CIPop/e29729d1c8d6871372f38c8cbcc43f93#file-client_ipv4_ipv6-c-L62